File: make-icons.py
#!/usr/bin/env python3
"""Regenerate the favicons and the link-preview image from the portrait.

The drawing in index.html is only a rough placeholder: hairD() rebuilds the
hair as smooth beziers at runtime, and the theme colours are picked at random
on load. So this renders the real page in headless Chrome and takes the
portrait from the *rendered* DOM — building straight from the source markup
gives you the angular, unsmoothed version.

Needs Chrome, rsvg-convert and ImageMagick:
    brew install librsvg imagemagick
"""

import re
import subprocess
import tempfile
from pathlib import Path

HERE = Path(__file__).parent
BG = "#fbfbfa"
INK = "#141414"
MUTED = "#a6a7ac"

# The square of the drawing's own coordinate space that frames the head.
HEAD_VIEWBOX = "332 110 404 404"

CHROME = "/Applications/Google Chrome.app/Contents/MacOS/Google Chrome"

# How the portrait is posed in the link preview: aim 0.5 looks straight out and
# lower turns the head to its left, toward the name. TRAVEL is the rig's own
# slider and caps how far it will turn, so it is raised past its 76 default to
# get a real three-quarter. The icons stay frontal at the defaults.
OG_AIM = 0.0
OG_TRAVEL = 140

# Pin the random theme to PAIRS[0] (black on cream), point the eyes straight
# ahead, let the springs settle, then freeze so the dump is a still frame.
# settle() must stay short: idle drift re-engages after ~5.5s, so a long settle
# walks the head back off-centre.
HARNESS_HEAD = "<script>Math.random=function(){return 0;};</script>\n"
HARNESS_TAIL = """
<script>
(function wait(){
  if (window.__erk) {
    var tr = __TRAVEL__;
    if (tr) {
      var el = document.querySelector('input[data-p="travel"]');
      if (el) { el.value = tr; el.dispatchEvent(new Event('input', {bubbles: true})); }
    }
    __erk.aim(__AIMX__, 0.5);
    __erk.settle(__SETTLE__);
    window.requestAnimationFrame = function(){ return 0; };
    document.documentElement.setAttribute('data-render-ready', '1');
  } else setTimeout(wait, 25);
})();
</script>
"""


def rendered_dom(aim_x=0.5, settle=1.0, travel=None) -> str:
    """index.html as the browser has it after the script has run and settled.

    aim_x is where the eyes are pointed across the stage: 0.5 looks straight
    out, lower values turn the head to its left. travel overrides the rig's
    TRAVEL slider, which caps how far the head will turn.
    """
    page = (HERE / "index.html").read_text()
    page = page.replace(
        '<script src="./support.js"></script>',
        HARNESS_HEAD + '<script src="./support.js"></script>', 1)
    tail = (HARNESS_TAIL.replace("__AIMX__", repr(float(aim_x)))
                        .replace("__SETTLE__", repr(float(settle)))
                        .replace("__TRAVEL__", repr(travel) if travel else "0"))

    # Must sit beside index.html so ./support.js still resolves.
    harness = HERE / ".render.html"
    harness.write_text(page + tail)
    try:
        dom = subprocess.run(
            [CHROME, "--headless", "--disable-gpu", "--no-sandbox",
             "--virtual-time-budget=8000", "--dump-dom", harness.resolve().as_uri()],
            check=True, capture_output=True, text=True,
        ).stdout
    finally:
        harness.unlink(missing_ok=True)

    if "data-render-ready" not in dom:
        raise SystemExit("the page never finished booting — nothing to capture")
    return dom


def portrait_markup(dom: str) -> str:
    """The contents of the portrait <svg>, as well-formed XML."""
    # The runtime adds its own attributes, so id="kSvg" is not necessarily first.
    marker = dom.index('id="kSvg"')
    start = dom.rindex("<svg", 0, marker)
    end = dom.index("</svg>", start) + len("</svg>")
    svg = dom[start:end]
    svg = svg.replace("will-change: transform;", "").replace("will-change:transform;", "")

    # Safety net: any unmatched </g> is fatal to an XML parser even though
    # browsers forgive it.
    tag = re.compile(r"<(/?)g(\s[^>]*?)?>")
    out, depth, pos = [], 0, 0
    for m in tag.finditer(svg):
        out.append(svg[pos:m.start()])
        pos = m.end()
        if m.group(1):
            if depth == 0:
                continue
            depth -= 1
        else:
            depth += 1
        out.append(m.group(0))
    out.append(svg[pos:])
    svg = "".join(out)

    inner = re.sub(r"^<svg\b[^>]*>", "", svg, count=1)
    if inner == svg:
        raise SystemExit("could not strip the <svg> wrapper")
    return inner[:inner.rindex("</svg>")]


def render(svg_text: str, width: int, height: int, out: Path) -> None:
    with tempfile.NamedTemporaryFile("w", suffix=".svg", delete=False) as f:
        f.write(svg_text)
        tmp = f.name
    subprocess.run(
        ["rsvg-convert", "-w", str(width), "-h", str(height), tmp, "-o", str(out)],
        check=True,
    )
    Path(tmp).unlink()


def content_viewbox(inner: str, pad: float = 1.2) -> str:
    """A square viewBox framing whatever the drawing actually covers.

    Turning the head moves it within the drawing's coordinate space, so a fixed
    crop clips the chin. Render on transparency, let ImageMagick report the
    trim box, and centre a square on it. 1 user unit renders as 1px here, so
    the box reads straight back as viewBox coordinates.
    """
    probe = (f'<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 960 720" '
             f'width="960" height="720">{inner}</svg>')
    with tempfile.TemporaryDirectory() as d:
        png = Path(d) / "probe.png"
        render(probe, 960, 720, png)
        box = subprocess.run(["magick", str(png), "-format", "%@", "info:"],
                             capture_output=True, text=True, check=True).stdout.strip()
    w, rest = box.split("x")
    h, x, y = rest.split("+")
    w, h, x, y = int(w), int(h), int(x), int(y)
    side = max(w, h) * pad
    cx, cy = x + w / 2, y + h / 2
    return f"{cx - side / 2:g} {cy - side / 2:g} {side:g} {side:g}"


def main() -> None:
    inner = portrait_markup(rendered_dom())
    x, y, size, _ = (float(v) for v in HEAD_VIEWBOX.split())

    def tile(radius: float) -> str:
        clip, open_g = "", "<g>"
        if radius:
            clip = (f'<defs><clipPath id="kTile"><rect x="{x:g}" y="{y:g}" '
                    f'width="{size:g}" height="{size:g}" rx="{radius:g}"/>'
                    f"</clipPath></defs>")
            open_g = '<g clip-path="url(#kTile)">'
        return (
            f'<svg xmlns="http://www.w3.org/2000/svg" viewBox="{HEAD_VIEWBOX}" '
            f'width="{size:g}" height="{size:g}">{clip}{open_g}'
            f'<rect x="{x:g}" y="{y:g}" width="{size:g}" height="{size:g}" '
            f'fill="#ffffff"/>{inner}</g></svg>'
        )

    rounded = tile(size * 0.22)
    (HERE / "icon.svg").write_text(rounded)

    # iOS masks apple-touch-icon itself and paints any transparency black, so
    # that one stays a square opaque tile.
    render(tile(0), 180, 180, HERE / "apple-touch-icon.png")

    with tempfile.TemporaryDirectory() as d:
        big = Path(d) / "icon512.png"
        render(rounded, 512, 512, big)
        subprocess.run(
            ["magick", str(big), "-define", "icon:auto-resize=48,32,16",
             str(HERE / "favicon.ico")],
            check=True,
        )

    # Link preview: three-quarter view, turned toward the name on the left.
    turned = portrait_markup(rendered_dom(aim_x=OG_AIM, settle=1.5, travel=OG_TRAVEL))
    og = f"""<svg xmlns="http://www.w3.org/2000/svg" width="1200" height="630" viewBox="0 0 1200 630">
<rect width="1200" height="630" fill="{BG}"/>
<svg x="655" y="55" width="520" height="520" viewBox="{content_viewbox(turned)}">{turned}</svg>
<text x="105" y="352" font-family="Helvetica Neue, Helvetica, Arial, sans-serif" font-size="112" font-weight="500" letter-spacing="-3" fill="{INK}">eric liu</text>
</svg>"""
    render(og, 1200, 630, HERE / "og.png")

    for name in ("icon.svg", "favicon.ico", "apple-touch-icon.png", "og.png"):
        print(f"{name:24} {(HERE / name).stat().st_size:>8,} bytes")


if __name__ == "__main__":
    main()

File: make-icons.mjs
#!/usr/bin/env node
// The favicons and the touch icon from the portrait, posed looking a little to
// its left, with nothing but Chrome (puppeteer from ../isola-quieta). The
// Python make-icons.py does the same with rsvg + ImageMagick and also writes
// og.png. Renders the real page (hairD() smooths the hair at runtime), aims
// the rig, waits for the springs, and screenshots the head square.
//   node make-icons.mjs        -> icon.svg, favicon.ico (48 px PNG inside), apple-touch-icon.png, og.png
// og.png (the link preview, 1200 x 630) is the real page at its resting angle - the name and the
// portrait as the site draws them - with the door and the icon row hidden.
import { createRequire } from 'node:module';
const puppeteer = (await import(createRequire(new URL('../isola-quieta/package.json', import.meta.url)).resolve('puppeteer'))).default;   // the viewer's puppeteer, whatever its layout
import { readFileSync, writeFileSync } from 'node:fs';
import { createServer } from 'node:http';
import { extname } from 'node:path';
const AIM = 0.28, TRAVEL = 120;   // 0.5 looks straight out; lower turns the head to its left
const AIM_OG = 0.44;              // the preview: the page's resting look, a slight turn (the three-quarter of the old og.png sat the lens on the nose)
const pageFor = (aim, extraCss = '') => readFileSync('index.html', 'utf8')
  .replace('<script src="./support.js"></script>', '<script>Math.random=function(){return 0;};</script>\n<script src="./support.js"></script>')
  + (extraCss ? `<style>${extraCss}</style>` : '')
  + `<script>(function wait(){ if (window.__erk) { var el = document.querySelector('input[data-p="travel"]'); if (el) { el.value = ${TRAVEL}; el.dispatchEvent(new Event('input', {bubbles: true})); }
     __erk.aim(${aim}, 0.5); __erk.settle(1.0); window.requestAnimationFrame = function(){ return 0; }; document.documentElement.setAttribute('data-render-ready', '1'); } else setTimeout(wait, 25); })();</script>`;
const page0 = pageFor(AIM);
let pageServed = page0;
const types = { '.html': 'text/html', '.js': 'text/javascript', '.css': 'text/css', '.svg': 'image/svg+xml', '.png': 'image/png' };
const server = createServer((req, res) => { const p = req.url.split('?')[0]; if (p === '/' || p === '/index.html') { res.setHeader('Content-Type', 'text/html'); res.end(pageServed); return; } try { res.setHeader('Content-Type', types[extname(p)] || 'application/octet-stream'); res.end(readFileSync('.' + p)); } catch (e) { res.statusCode = 404; res.end(); } });
await new Promise(r => server.listen(0, '127.0.0.1', r));
const url = `http://127.0.0.1:${server.address().port}/`;
const browser = await puppeteer.launch({ headless: true });
const page = await browser.newPage(); await page.setViewport({ width: 1440, height: 900, deviceScaleFactor: 1 });
await page.goto(url, { waitUntil: 'load' });
await page.waitForFunction(() => document.documentElement.getAttribute('data-render-ready') === '1', { timeout: 30000 });
await new Promise(r => setTimeout(r, 300));
// the rendered portrait, as the Python does it: the svg's inner markup, will-change stripped
const inner = await page.evaluate(() => { const svg = document.getElementById('kSvg'); return svg.innerHTML.replace(/will-change:\s*transform;?/g, ''); });
const HEAD = '332 110 404 404';
const icon = `<svg xmlns="http://www.w3.org/2000/svg" viewBox="${HEAD}" width="404" height="404"><defs><clipPath id="kTile"><rect x="332" y="110" width="404" height="404" rx="88.88"/></clipPath></defs><g clip-path="url(#kTile)"><rect x="332" y="110" width="404" height="404" fill="#ffffff"/>${inner}</g></svg>`;
writeFileSync('icon.svg', icon);
// raster: a page that shows the svg at the wanted size, screenshotted
const shot = async (size) => { const p = await browser.newPage(); await p.setViewport({ width: size, height: size, deviceScaleFactor: 1 });
  await p.setContent(`<!doctype html><html><body style="margin:0;background:transparent">${icon.replace('width="404" height="404"', `width="${size}" height="${size}"`)}</body></html>`);
  const buf = await p.screenshot({ type: 'png', omitBackground: true, clip: { x: 0, y: 0, width: size, height: size } }); await p.close(); return buf; };
writeFileSync('apple-touch-icon.png', await shot(180));
const png48 = await shot(48);
// favicon.ico: one 48 px PNG entry (every current browser reads PNG-in-ICO)
const ico = Buffer.alloc(6 + 16); ico.writeUInt16LE(0, 0); ico.writeUInt16LE(1, 2); ico.writeUInt16LE(1, 4);
ico[6] = 48; ico[7] = 48; ico[8] = 0; ico[9] = 0; ico.writeUInt16LE(1, 10); ico.writeUInt16LE(32, 12); ico.writeUInt32LE(png48.length, 14); ico.writeUInt32LE(22, 18);
writeFileSync('favicon.ico', Buffer.concat([ico, png48]));
// og.png: the page itself at 1200 x 630, resting, the door and the icon row hidden, the links under the name too
// (the site's runtime rebuilds the page from its template, so the styling goes in after the render:
// the door, the icon row and the links hidden, the portrait's box enlarged so the head fills the card)
pageServed = pageFor(AIM_OG);
const og = await browser.newPage(); await og.setViewport({ width: 1200, height: 630, deviceScaleFactor: 1 });
await og.goto(url, { waitUntil: 'load' });
await og.waitForFunction(() => document.documentElement.getAttribute('data-render-ready') === '1', { timeout: 30000 });
await og.addStyleTag({ content: '#kNav, #kPanel { display:none !important; } #kCopy > *:not(h1) { display:none !important; } #kStage { --pw: 1040px !important; --mr: -150px !important; } #kCopy { left: 8% !important; } #kCopy h1 { font-size: 104px !important; }' });
await new Promise(r => setTimeout(r, 400));
writeFileSync('og.png', await og.screenshot({ type: 'png', clip: { x: 0, y: 0, width: 1200, height: 630 } }));
await og.close();
await browser.close(); server.close();
console.log('icons: icon.svg', icon.length, 'B; apple-touch-icon.png 180 px; favicon.ico 48 px png; aim', AIM, '; og.png 1200x630 at aim', AIM_OG);
